python/tvm/testing.py

```diff
@@ -285,4 +288,184 @@ def _check_forward(constraints1, constraints2, varmap, backvarmap):
+
+
+def requires_cuda(f):
+    """Mark a test as requiring the CUDA runtime.
+
+    This also marks the test as requiring a gpu.
+
+    Parameters
+    ----------
+    f : function
+        Function to mark
+    """
+    return pytest.mark.cuda(
+        pytest.mark.skipif(
+            not tvm.runtime.enabled("cuda"), reason="CUDA support not enabled"
+        )(requires_gpu(f))
+    )
+
+
+def requires_opencl(f):
+    """Mark a test as requiring the OpenCL runtime.
+
+    This also marks the test as requiring a gpu.
+
+    Parameters
+    ----------
+    f : function
+        Function to mark
+    """
+    return pytest.mark.opencl(
+        pytest.mark.skipif(
+            not tvm.runtime.enabled("opencl"), reason="OpenCL support not enabled"
+        )(requires_gpu(f))
+    )
+
+
+def requires_rocm(f):
+    """Mark a test as requiring the rocm runtime.
+
+    This also marks the test as requiring a gpu.
+
+    Parameters
+    ----------
+    f : function
+        Function to mark
+    """
+    return pytest.mark.rocm(
+        pytest.mark.skipif(
+            not tvm.runtime.enabled("rocm"), reason="rocm support not enabled"
+        )(requires_gpu(f))
+    )
+
+
+def requires_tensorcore(f):
+    """Mark a test as requiring a tensorcore to run.
+
+    Tests with this mark will not be run unless a tensorcore is present.
+
+    Parameters
+    ----------
+    f : function
+        Function to mark
+    """
+    return pytest.mark.tensorcore(
+        pytest.mark.skipif(
+            not tvm.gpu().exist or not nvcc.have_tensorcore(tvm.gpu(0).compute_version),
+            reason="No tensorcore present",
+        )(f)
+    )
+
+
+def parametrize_devices(f):
+    """Parametrize a test over all enabled devices.
+
+    Parameters
+    ----------
+    f : function
+        Function to parametrize. Must be of the form `def test_xxxxxxxxx(device, ctx)`:,
+        where `xxxxxxxxx` is any name.
+
+    Example
+    -------
+    >>> @tvm.testing.parametrize
+    >>> def test_mytest(device, ctx):
+    >>>     ...  # do something
+    """
+    return pytest.mark.parametrize("device,ctx", enabled_devices())(uses_gpu(f))
+
+
+def _get_backends():
+    backend_str = os.environ.get("TVM_TEST_DEVICES", "")
+    if len(backend_str) == 0:
+        backend_str = DEFAULT_TEST_DEVICES
+    backends = {
+        dev
+        for dev in backend_str.split(";")
+        if len(dev) > 0 and tvm.context(dev, 0).exist and tvm.runtime.enabled(dev)
+    }
+    if len(backends) == 0:
+        logging.warning(
+            "None of the following backends are supported by this build of TVM: %s."
+            "Try setting TVM_TEST_DEVICES to a supported backend. Defaulting to llvm.",
+            backend_str
+        )
+        return {"llvm"}
+    return backends
+
+
+DEFAULT_TEST_DEVICES = (
+    "llvm;cuda;opencl;metal;rocm;vulkan;nvptx;"
+    "llvm -device=arm_cpu;opencl -device=mali,aocl_sw_emu"
+)
+TEST_DEVICES = _get_backends()
+
+
+def device_enabled(device):
```

Comment:
The `device` here is really a target.

Reply:
It seems like I'm confusing device and target here. I'll rename these functions to `target_enabled` and `enabled_targets`.

Reply:
"Target" is what we're compiling for, an abstraction of a machine with a specific ISA, etc.  A "device" is a concrete piece of hardware that is an example of the target.
What you're checking is the presence of devices, and also whether their corresponding targets are enabled (this is the `runtime.enabled` part).  In that sense, your function names are correct.  I was referring to the name of the parameter (sorry, it wasn't clear).